File: python/chapter03/generic_procedures.py
import itertools

from chapter03.trie import Trie


def generic_procedure_constructor(dispatch_store_maker):
    def generic_procedure(name, arity, default_handler):
        metadata = Metadata(name, arity, dispatch_store_maker(),
                            default_handler or error_generic_procedure_handler(name))

        def the_generic_procedure(*args):
            return generic_procedure_dispatch(metadata, args)

        set_generic_procedure_metadata(the_generic_procedure, metadata)
        return the_generic_procedure

    return generic_procedure


def error_generic_procedure_handler(name):
    def handler(*args):
        raise TypeError(f"{name} not applicable to {args}")

    return handler


def generic_procedure_dispatch(metadata, args):
    handler = get_generic_procedure_handler(metadata, args)
    return handler(*args)


def get_generic_procedure_handler(metadata, args):
    return metadata.getter(args) or metadata.default_getter()


def define_generic_procedure_handler(generic_procedure, applicability, handler):
    generic_procedure_metadata(generic_procedure).dispatch_store.add_handler(applicability, handler)


class SimpleDispatchStore:
    def __init__(self):
        self.rules = {}
        self.default_handler = None

    def get_handler(self, args):
        for predicates, handler in self.rules.items():
            if predicates_match(predicates, args):
                return handler
        return None

    def add_handler(self, applicability, handler):
        for predicates in applicability:
            self.rules[predicates] = handler

    def get_default_handler(self):
        return self.default_handler

    def set_default_handler(self, handler):
        self.default_handler = handler


class TrieDispatchStore(SimpleDispatchStore):
    def __init__(self):
        super().__init__()
        self.trie = Trie()

    def get_handler(self, args):
        return self.trie.get_a_value(args)

    def add_handler(self, applicability, handler):
        super().add_handler(applicability, handler)
        for path in applicability:
            self.trie.set_path_value(path, handler)


def cache_wrap_dispatch_store(DispatchStore):
    class WrappedDispatchStore(DispatchStore):
        def __init__(self):
            super().__init__()
            self.cache = {}

        def get_handler(self, args):
            types = tuple([type(a) for a in args])
            if types in self.cache:
                return self.cache[types]
            else:
                h = super().get_handler(args)
                self.cache[types] = h
                return h

    return WrappedDispatchStore


simple_generic_procedure = generic_procedure_constructor(SimpleDispatchStore)


class Metadata:
    def __init__(self, name, arity, dispatch_store, default_handler):
        dispatch_store.set_default_handler(default_handler)
        self.name = name
        self.arity = arity
        self.dispatch_store = dispatch_store
        self.getter = dispatch_store.get_handler
        self.default_getter = dispatch_store.get_default_handler


GENERIC_PROCEDURE_METADATA = {}


def set_generic_procedure_metadata(procedure, metadata):
    GENERIC_PROCEDURE_METADATA[procedure] = metadata


def generic_procedure_metadata(procedure):
    return GENERIC_PROCEDURE_METADATA[procedure]


def predicates_match(predicates, args):
    return all(pred(arg) for pred, arg in zip(predicates, args))


def all_args(arity, predicate):
    return [(predicate,) * arity]


def any_arg(arity, predicate, base_predicate):
    return [(predicate,) * arity] + list(itertools.permutations((predicate, base_predicate), arity))


def match_args(*predicates):
    return [predicates]


if __name__ == '__main__':
    def is_number(obj):
        PREDICATE_COUNTS[is_number] += 1
        return isinstance(obj, int) or isinstance(obj, float)


    def is_symbolic(obj):
        PREDICATE_COUNTS[is_symbolic] += 1
        return isinstance(obj, str)


    PREDICATE_COUNTS = None


    def reset_predicate_counts():
        global PREDICATE_COUNTS
        PREDICATE_COUNTS = {is_number: 0, is_symbolic: 0}


    reset_predicate_counts()

    generic_procedure = generic_procedure_constructor(cache_wrap_dispatch_store(TrieDispatchStore))

    plus = generic_procedure("plus", 2, None)
    define_generic_procedure_handler(plus, all_args(2, is_number), lambda a, b: a + b)
    define_generic_procedure_handler(plus, any_arg(2, is_symbolic, is_number), lambda a, b: f"({a} + {b})")

    minus = generic_procedure("minus", 2, None)
    define_generic_procedure_handler(minus, all_args(2, is_number), lambda a, b: a - b)
    define_generic_procedure_handler(minus, any_arg(2, is_symbolic, is_number), lambda a, b: f"({a} - {b})")

    print(plus(1, 2))
    print(plus(1, "b"))
    print(plus("a", 2))
    print(plus("a", "b"))

    reset_predicate_counts()


    def fib(n):
        if n <= 2:
            return n
        else:
            return plus(fib(minus(n, 1)), fib(minus(n, 2)))


    print(fib(20))
    print(PREDICATE_COUNTS)

File: rust/src/chapter03/generic_procedures.rs
use lazy_static::lazy_static;
use std::any::Any;
use std::collections::HashMap;
use std::hash::{Hash, Hasher};
use std::sync::{Arc, RwLock, Weak};

macro_rules! obj {
    (mut $trait:path) => { Arc<RwLock<dyn $trait + Send + Sync>> };
    ($trait:path) => { Arc<dyn $trait + Send + Sync> };

    (new mut $value:expr) => { Arc::new(RwLock::new($value)) };
}

pub trait Object: Any + std::fmt::Debug {
    fn as_any(&self) -> &dyn Any;
}
impl<T: Any + std::fmt::Debug> Object for T {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

type GenericFn = Arc<dyn 'static + Sync + Send + Fn(GenericArgs<'_, '_>) -> GenericResult>;
type GenericResult = Result<Arc<dyn Object>, Error>;
type GenericArgs<'a, 'o> = &'a [&'o dyn Object];

type Handler = GenericFn;

type Applicability = Vec<Vec<Predicate>>;
type PredicateFn = fn(&dyn Object) -> bool;

#[derive(Copy, Clone)]
pub struct Predicate(PredicateFn);

impl From<PredicateFn> for Predicate {
    fn from(func: PredicateFn) -> Self {
        Predicate(func)
    }
}

impl PartialEq for Predicate {
    fn eq(&self, other: &Self) -> bool {
        std::ptr::eq(&self.0, &other.0)
    }
}

impl Predicate {
    fn check(&self, obj: &dyn Object) -> bool {
        (self.0)(obj)
    }
}

type Error = Arc<dyn ToString>;

pub fn make_generic_procedure_constructor<T: DispatchStore>(
    dispatch_store_maker: impl 'static + Fn() -> T,
) -> impl Fn(&str, Option<Handler>) -> GenericFn {
    let constructor = move |name: &str, default_handler| {
        let metadata =
            ConcreteMetadata::new(name.to_string(), dispatch_store_maker(), default_handler);
        let metadata: obj![mut Metadata] = obj!(new mut metadata);

        let the_generic_procedure: GenericFn = {
            let metadata = metadata.clone();
            Arc::new(move |args| metadata.read().unwrap().dispatch(args))
        };

        set_generic_metadata(&the_generic_procedure, metadata);

        the_generic_procedure
    };
    constructor
}

pub fn define_generic_procedure_handler(
    generic_procedure: &GenericFn,
    applicability: Applicability,
    handler: impl 'static + Sync + Send + Fn(GenericArgs<'_, '_>) -> GenericResult,
) {
    let handler = Arc::new(handler);
    let key = MetaKey(Arc::downgrade(generic_procedure));
    METADATA.write().unwrap()[&key]
        .write()
        .unwrap()
        .add_handler(applicability, handler);
}

trait Metadata {
    fn add_handler(&mut self, applicability: Applicability, handler: Handler);
    fn dispatch(&self, args: GenericArgs) -> GenericResult;
}

#[derive(Clone)]
struct ConcreteMetadata<T: DispatchStore> {
    dispatch_store: T,
    default_handler: Handler,
}

impl<T: DispatchStore> ConcreteMetadata<T> {
    pub fn new(name: String, dispatch_store: T, default_handler: Option<Handler>) -> Self {
        let default_handler = default_handler.unwrap_or_else(|| {
            Arc::new(move |args| {
                Err(Arc::new(format!(
                    "Cannot apply {} to ({})",
                    name,
                    args.iter()
                        .map(|a| format!("{:?}", a))
                        .collect::<Vec<_>>()
                        .join(", ")
                )))
            })
        });

        ConcreteMetadata {
            dispatch_store,
            default_handler,
        }
    }

    pub fn default_handler(&self, args: GenericArgs) -> GenericResult {
        (self.default_handler)(args)
    }
}

impl<T: DispatchStore> Metadata for ConcreteMetadata<T> {
    fn add_handler(&mut self, applicability: Applicability, handler: Handler) {
        self.dispatch_store.add_handler(applicability, handler)
    }

    fn dispatch(&self, args: GenericArgs) -> GenericResult {
        match self.dispatch_store.get_handler(args) {
            Some(handler) => handler(args),
            None => self.default_handler(args),
        }
    }
}

fn set_generic_metadata(genfn: &GenericFn, metadata: obj![mut Metadata]) {
    let key = MetaKey(Arc::downgrade(genfn));
    METADATA.write().unwrap().insert(key, metadata);
}

lazy_static! {
    static ref METADATA: RwLock<HashMap<MetaKey, obj![mut Metadata]>> = RwLock::new(HashMap::new());
}

struct MetaKey(Weak<dyn Sync + Send + Fn(GenericArgs<'_, '_>) -> GenericResult>);

impl MetaKey {
    fn as_ptr(&self) -> *const (dyn Sync + Send + Fn(GenericArgs<'_, '_>) -> GenericResult) {
        Weak::as_ptr(&self.0)
    }
}

impl Eq for MetaKey {}
impl PartialEq for MetaKey {
    fn eq(&self, other: &Self) -> bool {
        std::ptr::eq(self.as_ptr(), other.as_ptr())
    }
}

impl Hash for MetaKey {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.as_ptr().hash(state)
    }
}

pub trait DispatchStore: 'static + Send + Sync {
    fn add_handler(&mut self, applicability: Applicability, handler: Handler);
    fn get_handler(&self, args: GenericArgs) -> Option<&Handler>;
}

struct SimpleDispatchStore {
    rules: Vec<(Vec<Predicate>, Handler)>,
}

impl SimpleDispatchStore {
    pub fn new() -> Self {
        SimpleDispatchStore { rules: vec![] }
    }
    fn predicates_match(predicates: &[Predicate], args: GenericArgs) -> bool {
        if args.len() != predicates.len() {
            return false;
        }

        predicates
            .iter()
            .zip(args)
            .all(|(pred, arg)| pred.check(*arg))
    }
}

impl DispatchStore for SimpleDispatchStore {
    fn add_handler(&mut self, applicability: Applicability, handler: Handler) {
        for rule in applicability {
            self.rules.push((rule, handler.clone()))
        }
    }

    fn get_handler(&self, args: GenericArgs) -> Option<&Handler> {
        for (predicates, handler) in &self.rules {
            if Self::predicates_match(predicates, args) {
                return Some(handler);
            }
        }
        None
    }
}

fn all_args(arity: usize, predicate: PredicateFn) -> Applicability {
    vec![vec![predicate.into(); arity]]
}

fn is_i64(obj: &dyn Object) -> bool {
    obj.as_any().downcast_ref::<i64>().is_some()
}

fn is_string(obj: &dyn Object) -> bool {
    obj.as_any().downcast_ref::<String>().is_some()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let generic_procedure = make_generic_procedure_constructor(SimpleDispatchStore::new);

        let plus = generic_procedure("plus", None);
        define_generic_procedure_handler(&plus, all_args(2, is_i64), |args| match args {
            [a, b] => {
                let a = (*a).as_any().downcast_ref::<i64>().unwrap();
                let b = (*b).as_any().downcast_ref::<i64>().unwrap();
                Ok(Arc::new(a + b))
            }
            _ => Err(Arc::new("wrong number of arguments")),
        });

        println!("{:?}", plus(&[&1i64, &2i64]).map_err(|e| e.to_string()));

        println!(
            "{:?}",
            plus(&[&"foo".to_string(), &"bar".to_string()]).map_err(|e| e.to_string())
        );

        define_generic_procedure_handler(&plus, all_args(2, is_string), |args| match args {
            [a, b] => {
                let a = (*a).as_any().downcast_ref::<String>().unwrap();
                let b = (*b).as_any().downcast_ref::<String>().unwrap();
                Ok(Arc::new(a.clone() + b))
            }
            _ => Err(Arc::new("wrong number of arguments")),
        });

        println!(
            "{:?}",
            plus(&[&"foo".to_string(), &"bar".to_string()]).map_err(|e| e.to_string())
        );
    }
}
